basic_bootloader/src/bootloader/mod.rs

```diff
@@ -400,4 +418,166 @@ where
+                tracer,
+            )?;
+        }
+
+        let native_resources_used_by_calls = native_resource_before_processing
+            .checked_sub(resources.native().as_u64())
+            .ok_or_else(|| internal_error!("Unexpected amount of native resources used"))?;
+
+        native_resource_used = native_resources_used_by_calls.saturating_add(native_resource_used);
+
+        Ok((interop_roots, native_resource_used))
+    }
+
+    /// Calls the L2 interop root storage contract to register a single interop root.
+    /// Returns updated resource limits after the contract call.
+    /// TODO: add_interop_roots_to_l2_interop_root_storage_batch should be used instead (see below)
+    fn add_interop_root_to_l2_interop_root_storage(
+        chain_id: u64,
+        block_or_batch_number: u64,
+        sides: &[Bytes32],
+        system: &mut System<S>,
+        system_functions: &mut HooksStorage<S, S::Allocator>,
+        memories: &mut RunnerMemoryBuffers,
+        resources: S::Resources,
+        tracer: &mut impl Tracer<S>,
+    ) -> Result<S::Resources, BootloaderSubsystemError>
+    where
+        S::IO: IOSubsystemExt,
+    {
+        let mut data = [0u8; 164];
+        // fb6200c6: function addInteropRoot(uint256 chainId, uint256 blockOrBatchNumber, bytes32[] calldata sides) external;
+        data[0..4].copy_from_slice(&[0xfb, 0x62, 0x00, 0xc6]);
+        data[28..36].copy_from_slice(&chain_id.to_be_bytes());
+        data[60..68].copy_from_slice(&block_or_batch_number.to_be_bytes());
+        data[96..100].copy_from_slice(&96u32.to_be_bytes());
+        data[128..132].copy_from_slice(&1u32.to_be_bytes());
+        data[132..164].copy_from_slice(&sides[0].as_u8_ref());
+
+        let res = Self::run_single_interaction(
+            system,
+            system_functions,
+            memories.reborrow(),
+            &data,
+            &BOOTLOADER_FORMAL_ADDRESS,
+            &L2_INTEROP_ROOT_STORAGE_ADDRESS,
+            resources,
+            &U256::ZERO,
+            true,
+            tracer,
+        )?;
+
+        match res.result {
+            CallResult::PreparationStepFailed => Err(internal_error!(
+                "Unexpected preparation failure in interop roots processing"
+            )
+            .into()), // Should never happen
+            CallResult::Failed { return_values } => Err(interface_error!(
+                BootloaderInterfaceError::FailedToSetInteropRoots
+            ))
+            .with_context(|| {
+                error_ctx! {
+                     "return_values" => debug_format(return_values),
+                }
+            }),
+            CallResult::Successful { return_values: _ } => Ok(res.resources_returned),
+        }
+    }
+
+    /// Calls the L2 interop root storage contract to register multiple interop roots in a single batch.
+    /// Returns updated resource limits after the contract call.
+    /// TODO: use after L2InteropRootStorage.sol supports corresponding functionality
```

Comment:
We should use this function if https://github.com/matter-labs/era-contracts/pull/1644 will be merged and deployed

It allows us to do just one call instead of 1 call per 1 root. Ideally we also want to wrap it as a formal system transaction. Otherwise it will emit events that will pollute the first tx in the block.